oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -83,7 +78,11 @@ pub struct Runtime {
 
     channels: channel::ChannelMapping,
 
-    nodes: RwLock<HashMap<NodeId, Node>>,
+    /// Runtime-specific state for each node instance.
+    node_infos: RwLock<HashMap<NodeId, NodeInfo>>,
+
+    /// Currently running node instances, so that [`Runtime::stop`] can terminate all of them.
+    node_instances: Mutex<HashMap<NodeId, Box<dyn crate::node::Node>>>,
```

Comment:
These two fields look like they should be covered by the same `Mutex`.

(In C++ this would probably be a base class that holds data (&approx; `NodeInfo`), but which still has pure virtual functions to be overridden (&approx; `node::Node`).  Not completely sure what the Rust equivalent would be &ndash; maybe to extend the `node::Node` trait to cover behaviour encompassed by what's in `NodeInfo`, and then each implementor of the trait has-a `NodeInfo` for that.)

Reply:
Having small granular locks for things that may need to be updated in a synchronized fashion implies a locking hierarchy, and there's nothing here that documents such a locking hierarchy.

The obvious example is if there's (say) a current chunk of code that adds a new Node, and which does so by first adding to `node_infos` then adds to `node_instances`.  If someone later adds different code that happens to add things to `node_instances` before `node_infos`, then there's a deadlock.

If all related state &ndash; here, everything that's indexed by `NodeId` &ndash; is under the same lock, then there's no trap for unwary maintainers in the future.

Reply:
But the point of this PR is exactly to decouple those two fields, so that they can be updated independently (no need to ever hold a lock over both). I have added a helper method `add_node_instance` to make things clearer now, and added more comments. PTAL.